cmd/{snap-seccomp,snap-confine-ns},osutil,interfaces/account_control: workaround unit test failures on Arch

bboozzoo opened the pull request:
There is no 'shadow' group in ArchLinux, what causes a failure in `cmd/snap-seccomp.TestCompile`. Try to workaround the problem by looking at owning gid of `/etc/shadow` and figuring out the corresponding group name based on this.

As a side effect, a helper for looking up group name by gid was added. The implementation is based on similar code in https://tip.golang.org/src/os/user/cgo_lookup_unix.go

@zyga can you take a look?

zyga commented on cmd/snap-seccomp/main_test.go:
```diff
@@ -350,8 +351,20 @@ func (s *snapSeccompSuite) TestUnrestricted(c *C) {
+	// /etc/shadow and use the gid that owns it
+	var shadowGid uint64
+	if gid, err := osutil.FindGid("shadow"); err != nil {
+		var stat syscall.Stat_t
```
Why not just `os.Stat`?

bboozzoo replied:
`os.Stat` does not carry any group information I'm afraid.

zyga commented on cmd/snap-seccomp/main_test.go:
```diff
@@ -350,8 +351,20 @@ func (s *snapSeccompSuite) TestUnrestricted(c *C) {
+	var shadowGid uint64
+	if gid, err := osutil.FindGid("shadow"); err != nil {
+		var stat syscall.Stat_t
+		err := syscall.Stat("/etc/shadow", &stat)
```
I think we could twist this test a little by just stating `/etc/shadow` unconditionally and using whatever the group there for the rest of the code below. Alternatively use something that works on Arch as well and just carry on.

The real problem is in the `account_control` interface which grants fchown to `u:root` and `g:shadow`. This will fail at runtime and needs to be fixed to grant the actual group owner of `/etc/shadow`

bboozzoo replied:
Agreed, skipping the fallback and doing `stat()` unconditionally will make the code easier. I'll amend the PR.

bboozzoo commented:
Added one last patch to workaround missing 'nogroup' and rebased on top of https://github.com/snapcore/snapd/commit/c6d06825490a863cd75aee256d86f371e7b13368

chipaca commented on osutil/group.go:
```diff
@@ -156,3 +184,28 @@ func FindGid(group string) (uint64, error) {
 	return strconv.ParseUint(group, 10, 64)
 }
+
+// FindGid returns the identifier of the given UNIX group name.
```
`FindGroup`

bboozzoo replied:
fixed

chipaca commented on osutil/group.go:
```diff
@@ -63,6 +63,34 @@ func lookupGroup(groupname string) (string, error) {
 }
 
+// Use implementation from:
+// https://golang.org/src/os/user/cgo_lookup_unix.go
```
could we take this one step further and pull in the `Group` type?

bboozzoo replied:
you mean this one? 
```
// from https://golang.org/src/os/user/user.go
type Group struct {
	Gid  string // group ID
	Name string // group name
}
```

chipaca replied:
yes

bboozzoo replied:
I pulled in `os/user.Group` as suggested. `FindGroupOwning()` was updated to return `*Group, error`, other calls were returning either gid or name, so I left those intact.

chipaca commented on osutil/group.go:
```diff
@@ -63,6 +63,34 @@ func lookupGroup(groupname string) (string, error) {
+	buf := alloc(groupBuffer)
+	defer buf.free()
+	err := retryWithBuffer(buf, func() syscall.Errno {
+		// mygetgrgid_r is a wrapper around getgrgid_r to
```
I suspect this comment does not apply

bboozzoo replied:
yup, thanks for spotting this

chipaca commented on osutil/group.go:
```diff
@@ -63,6 +63,34 @@ func lookupGroup(groupname string) (string, error) {
+
+	})
+	if err != nil {
+		return "", fmt.Errorf("group: lookup groupid %d: %v", gid, err)
```
`"cannot lookup group %d: %v"` or somesuch (key being the starting with "cannot")

chipaca commented on osutil/group.go:
```diff
@@ -63,6 +63,34 @@ func lookupGroup(groupname string) (string, error) {
+		return "", fmt.Errorf("group: lookup groupid %d: %v", gid, err)
+	}
+	if result == nil {
+		return "", fmt.Errorf("group: unknown group %d", gid)
```
errors should usually start with a `cannot <do this thing you asked me to do>`

bboozzoo replied:
updated both messages

mvo5 commented on cmd/snap-seccomp/main_test.go:
```diff
@@ -350,8 +350,9 @@ func (s *snapSeccompSuite) TestUnrestricted(c *C) {
-	shadowGid, err := osutil.FindGid("shadow")
+	// The 'shadow' group is different in different distributions, instead
+	// of second guessing group name, look at the owner of /etc/shadow
+	shadowGid, shadowGroup, err := osutil.FindGroupOwning("/etc/shadow")
```
Nice!

mvo5 commented on osutil/group.go:
```diff
@@ -156,3 +184,28 @@ func FindGid(group string) (uint64, error) {
 }
+
+// FindGid returns the identifier of the given UNIX group name.
+func FindGroup(gid uint64) (string, error) {
```
Can we write tests for this? A simple smoke test based on the local user or something?

zyga replied:
Ha, the reason this started was that there's just nothing sane about groups across systems. 

Still +1 on a test.

bboozzoo replied:
Added some tests to cover find group ID/name and group owning a particular file.

mvo5 commented on osutil/group.go:
```diff
@@ -156,3 +184,28 @@ func FindGid(group string) (uint64, error) {
+}
+
+// FindGroupOwning obtains UNIX group ID and name owning file `path`.
+func FindGroupOwning(path string) (uint64, string, error) {
```
And this as well? Again, something simple is enough for me. E.g. create a file, ensure its the group  of the user running the test and call this helper or something like this?

bboozzoo commented:
`tests/main/static` fails, because `snapd` can no longer be built with `CGO_ENABLED=0`. This is perhaps something worth discussing on the forums/IRC. Disabling `cgo` only works as long as you don't need `cgo`. In our case, we actually need it because of calls to `getgrgid_r`. When building with `cgo` disabled, Go toolchain was so nice to silently skip building `osutil/group.go` without a single word of warning only to producing uninformative `interfaces/builtin/account_control.go:84:16: undefined: osutil.FindGroupOwning`.

I will update the test to use `-ldflags '-extldflags "-static"'`, but it's only half of solution. We need to decide which binaries **must** be built statically and which do not. As I understand, `snap-seccomp`, `snap-exec`, `snap-update-ns` are candidates. Out of these 3, `snap-seccomp` is a problem on Arch, as libseccomp is only avaialble as `*.so`. Given this, maybe `tests/main/static` doesn't even make sense.

Thoughts?

zyga commented:
@bboozzoo to expand on the which binaries need to be statically linked. This is a bit more subtle and complex. The Ubuntu 16.04 build is special as it is used to construct the core snap which is participating in re-execution in many places. 

On Arch we don't really need to link anything statically until we are ready to use base snaps where the base snap transition in the execution environment will pose a problem. At that time we should be able to use snapd from core so again, we don't really strongly need that.

chipaca commented on osutil/group_test.go:
```diff
@@ -61,9 +63,9 @@ func (s *groupSuite) TestSelfOwnedFile(c *C) {
-
-	c.Check(group.Gid, Equals, self.Gid)
+	if c.Check(err, IsNil) {
+		c.Check(group.Gid, Equals, self.Gid)
```
this is usually written
```golang
c.Assert(err, IsNil)
c.Check(...)
```

chipaca commented on osutil/group_test.go:
```diff
@@ -0,0 +1,75 @@
+
+func (s *groupSuite) TestSelfOwnedFile(c *C) {
+	self, err := RealUser()
+	c.Check(err, IsNil)
```
in general you'd use `Assert` on errors, unless the error doesn't affect the rest of the test; here I think it very much does

chipaca commented on osutil/group_test.go:
```diff
@@ -0,0 +1,75 @@
+	defer os.Remove(name)
+
+	group, err := FindGroupOwning(name)
+	if c.Check(err, IsNil) {
```
and here, `c.Assert(err, IsNil)` and you can do away with the `if`

bboozzoo commented:
Do we need another round of reviews or is this good to be merged now?

zyga commented:
I'll do second review shortly (just woke up)

mvo5 commented on cmd/snap-update-ns/entry_test.go:
```diff
@@ -106,9 +108,21 @@ func (s *entrySuite) TestXSnapdGid(c *C) {
-	e = &mount.Entry{Options: []string{"x-snapd.gid=nogroup"}}
+	var nogroup string
+	var nogroupGid uint64
+	// try to find a suitable 'nogroup' like group
```
(nitpick) Maybe expand the comment a tiny bit? Something like: `try to find a suitable 'nogroup' like group, e.g. arch uses "nobody"`

mvo5 commented on interfaces/builtin/account_control.go:
```diff
@@ -56,26 +64,55 @@ capability fsetid,
+func (iface *accountControlInterface) SecCompConnectedPlug(spec *seccomp.Specification, plug *interfaces.Plug, Attrs map[string]interface{}, slot *interfaces.Slot, slotAttrs map[string]interface{}) error {
+	if snippet, err := makeAccountControlSecCompSnippet(); err != nil {
+		return err
+	} else {
```
(nitpick) we don't need the `} else {` here our convention is like: foo, err := thing()\n if err != nil {return err}\nuse(thing). So something like the lines below will be more idiomatic for our code base:
```
        snippet, err := makeAccountControlSecCompSnippet()
        if err != nil {
		return err
	}
	spec.AddSnippet(snippet)
	return nil
```
is fine.

mvo5 commented on osutil/group_test.go:
```diff
@@ -0,0 +1,74 @@
+type groupSuite struct {
+}
+
+var _ = Suite(&groupSuite{})
```
Thanks for adding those tests!

mvo5 commented on osutil/group_test.go:
```diff
@@ -0,0 +1,74 @@
+ *
+ */
+
+package osutil
```
(nitpick) we use a slightly different pattern when writing (most) unit tests. The package is called `foo_test` instead of `foo` which forces to use only the public interfaces of the thing that gets tested. This is similar to what the go upstream source is using. If internal access is needed we use a `export_test.go` file that is part of package "foo" to only export things for tests. For this particular test changes should be minimal as it is only testing public functions. So just renaming the package from `osutil` to `osutil_test` should be enough (plus import of osutil and prefixing the methods).

mvo5 commented on osutil/group_test.go:
```diff
@@ -0,0 +1,74 @@
+	. "gopkg.in/check.v1"
+)
+
+type groupSuite struct {
```
(nitpick^2) empty structs are mostly written `type groupSuite struct {}`, i.e. no extra newline. We don't always follow that rule but mostly :)

bboozzoo commented:
Merge?

chipaca commented:
once green, go for it

bboozzoo commented:
Interesting, travis failed with:
```
2017-11-08 12:33:57 Failed tasks: 1
    - linode:ubuntu-core-16-64:tests/main/snap-auto-mount
2017-11-08 12:33:57 Failed task prepare: 1
    - linode:fedora-25-64:tests/main/classic-custom-device-reg
2017-11-08 12:33:57 Failed task restore: 1
    - linode:fedora-25-64:tests/main/classic-custom-device-reg
```

linode:fedora-25-64:tests/main/classic-custom-device-reg error is:
```
error: cannot pack "/home/gopath/src/github.com/snapcore/snapd/tests/lib/snaps/classic-gadget": exit status 1
```

linode:ubuntu-core-16-64:tests/main/snap-auto-mount error:
```
+ echo 'The auto-mount magic has given us the assertion'
The auto-mount magic has given us the assertion
+ MATCH 'name: test-store'
+ snap known account-key
error: pattern not found, got:
```

bboozzoo commented:
Closing this for now. As discussed in the meetng, we'll try to kill most of `cgo` code, meaning we'll drop `g:{{group}}` filter form account control seccomp template followed by dropping of all unnecessary group query code.
